File: src/com/mantono/ghapic/WorkManager.java
package com.mantono.ghapic;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalUnit;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class WorkManager extends ThreadPoolExecutor
{
	public static final int HOURLY_RATE = 5000;
	public static final int MINUTE_SEARCH_RATE = 30;
	private final BlockingQueue<Runnable> workQueue;
	private Instant limitResetTime = Instant.now();
	private Instant limitSearchResetTime = Instant.now();
	private int remainingRequests = HOURLY_RATE;
	private int remainingSearchRequests = MINUTE_SEARCH_RATE;

	public WorkManager(final int corePoolSize, final int maximumPoolSize, final long keepAliveTime, final TimeUnit unit, BlockingQueue<Runnable> workQueue)
	{
		super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
		this.workQueue = workQueue;
	}
	
	@Override
	protected void beforeExecute(Thread t, Runnable r)
	{
		super.beforeExecute(t, r);
		sleep();
		while(limitIsReached() || searchLimitIsReached())
			Thread.yield();
	}
	
	@Override
	protected void afterExecute(Runnable r, Throwable t)
	{
		super.afterExecute(r, t);
		try
		{
			if(!(r instanceof Future<?>))
				return;
			
			Future<?> future = (Future<?>) r;
			Object obj = future.get();
			
			if(!(obj instanceof Response))
				return;
			
			Response response = (Response) obj;
			
			final int rateLimit = Integer.parseInt(response.getHeaderField("X-RateLimit-Limit"));
			
			final int remaining = Integer.parseInt(response.getHeaderField("X-RateLimit-Remaining"));
			final String resetTime = response.getHeaderField("X-RateLimit-Reset");
			final long time = Long.parseLong(resetTime);
			
			if(rateLimit == HOURLY_RATE)
			{
				this.remainingRequests = remaining;
				setResetTime(time);
			}
			else if(rateLimit == MINUTE_SEARCH_RATE)
			{
				this.remainingSearchRequests = remaining;
				setSearchResetTime(time);
			}
		
		}
		catch(InterruptedException e)
		{
			e.printStackTrace();
		}
		catch(ExecutionException e)
		{
			e.printStackTrace();
		}
	}

	private long sleepTime()
	{
		return 5300/(remainingResourceRequests() - pendingRequests());
	}
	
	private long sleep()
	{
		final long sleep = sleepTime();
		try
		{
			if(sleep > 0)
				Thread.sleep(sleep);
		}
		catch(InterruptedException e)
		{
			e.printStackTrace();
		}
		return sleep;
	}
	
	public int consumedRequests()
	{
		return HOURLY_RATE - remainingRequests;
	}
	
	public int consumedSearchRequests()
	{
		return MINUTE_SEARCH_RATE - remainingSearchRequests;
	}
	
	public int remainingResourceRequests()
	{
		return remainingRequests;
	}
	
	public int remainingSearchRequests()
	{
		return remainingSearchRequests;
	}
	
	public int pendingRequests()
	{
		return workQueue.size();
	}
	
	public boolean limitIsReached()
	{
		return remainingResourceRequests() - pendingRequests() < 20;
	}
	
	public boolean searchLimitIsReached()
	{
		return remainingSearchRequests() < 1;
	}
	
	/**
	 * 
	 * @return the amount of seconds until the rate limit resets.
	 */
	public long timeUntilReset()
	{
		return Instant.now().until(limitResetTime, (TemporalUnit) ChronoUnit.SECONDS);
	}
	
	/**
	 * 
	 * @return the amount of seconds until the rate limit for searches resets.
	 */
	public long timeUntilSearchReset()
	{
		return Instant.now().until(limitSearchResetTime, (TemporalUnit) ChronoUnit.SECONDS);
	}
	
	private void setResetTime(long time)
	{
		this.limitResetTime = Instant.ofEpochSecond(time);
	}
	
	private void setSearchResetTime(long time)
	{
		this.limitSearchResetTime = Instant.ofEpochSecond(time);
	}
}

File: src/main/kotlin/com/mantono/ghapic/WorkManager.kt
package com.mantono.ghapic

import java.time.Instant
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalUnit
import java.util.concurrent.*

const val HOURLY_RATE: Int = 5000
const val MINUTE_SEARCH_RATE: Int = 30

class WorkManager(corePoolSize: Int = 4,
                  maximumPoolSize: Int = 2 * corePoolSize,
                  keepAliveTime: Long = 5000L,
                  unit: TimeUnit = TimeUnit.MILLISECONDS,
                  private val workQueue: BlockingQueue<Runnable>) : ThreadPoolExecutor(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue)
{
	private var resetTime: Instant = Instant.now()
	private var searchResetTime: Instant = Instant.now()
	private var remainingRequests: Int = HOURLY_RATE
		get
	private var remainingSearchRequests: Int = MINUTE_SEARCH_RATE
		get

	override fun beforeExecute(t: Thread, r: Runnable)
	{
		super.beforeExecute(t, r)
		sleep()
		while(limitIsReached() || searchLimitIsReached())
			Thread.yield()
	}

	override fun afterExecute(r: Runnable, t: Throwable)
	{
		super.afterExecute(r, t)

		if (r !is Future<*>)
			return

		val future = r as Future<*>?
		val response: Response = future!!.get() as? Response ?: return

		val rateLimit = Integer.parseInt(response["X-RateLimit-Limit"])

		val remaining = Integer.parseInt(response["X-RateLimit-Remaining"])
		val resetTime = response["X-RateLimit-Reset"]
		val time = java.lang.Long.parseLong(resetTime)

		if (rateLimit == HOURLY_RATE)
		{
			this.remainingRequests = remaining
			setResetTime(time)
		}
		else if (rateLimit == MINUTE_SEARCH_RATE)
		{
			this.remainingSearchRequests = remaining
			setSearchResetTime(time)
		}


	}

	private fun sleepTime(): Long = (5300 / (remainingRequests - pendingRequests())).toLong()

	private fun sleep(): Long
	{
		sleepTime().run {
			if (this > 0)
				Thread.sleep(this)
			return this
		}
	}

	fun consumedRequests(): Int = HOURLY_RATE - remainingRequests
	fun consumedSearchRequests(): Int = MINUTE_SEARCH_RATE - remainingSearchRequests
	fun pendingRequests(): Int = workQueue.size
	fun limitIsReached(): Boolean = remainingRequests - pendingRequests() < 20
	fun searchLimitIsReached(): Boolean = remainingSearchRequests < 1

	/**
	 *
	 * @return the amount of seconds until the rate limit resets.
	 */
	fun timeUntilReset(): Long = Instant.now().until(resetTime, ChronoUnit.SECONDS as TemporalUnit)

	/**
	 *
	 * @return the amount of seconds until the rate limit for searches resets.
	 */
	fun timeUntilSearchReset(): Long = Instant.now().until(searchResetTime, ChronoUnit.SECONDS as TemporalUnit)

	private fun setResetTime(time: Long)
	{
		this.resetTime = Instant.ofEpochSecond(time)
	}

	private fun setSearchResetTime(time: Long)
	{
		this.searchResetTime = Instant.ofEpochSecond(time)
	}
}